### DISCUSSION TITLE
How to get hold of MauiAppBuilder instance?

### DISCUSSION THREAD

**Discussion by ataparia** [dotnet/maui#10382] [General] [answered]
Hi Team,
I need to get hold of MauiAppBuilder instance in a view model. To be precise, I need the "Services" property that comes off the MauiAppBuilder class so that I can get to a specific service as this service is not injected through the constructor. Is there any way to get hold of this property in the viewmodel classes?

**Comment by Ghevi** [comment]
I think you can inject `IServiceProvider` and do
`var myService = serviceProvider.GetRequiredService<IMyService>();`

 Or you can inject `IServiceScopeFactory` and do
` using var scope = scopeFactory.CreateScope();`
 `var myService = scope.ServiceProvider.GetRequiredService<IMyService>();`

Does this help you?

**Reply by ataparia** [reply]
Thanks @Ghevi. That doesnt work I guess. I get runtime exception when I put IServiceProvider in the constructor of the MainPage.xaml.cs. I dont know if its a framework limitation where we cannot use constructor injection in MainPage.xaml.cs

**Reply by Ghevi** [reply]
Ahh I see, you also have to register MainPage and AppShell in `MauiProgram.cs` and then set the MainPage property in the `AppShell.cs`, see the [repository](https://github.com/Ghevi/maui-dependency-injection) I've uploaded, then it works.
I'm not sure if this is the best way tho.

You can also check this [video](https://www.youtube.com/watch?v=xx1mve2AQr4) that explains DI better.

**Comment by ataparia** [comment, marked as answer]
Thanks a ton @Ghevi. Seems like I have to register MainPage in the MauiProgram. It works perfectly. Thanks again for your help.

**Reply by Ghevi** [reply]
You're welcome, have a good day :)
